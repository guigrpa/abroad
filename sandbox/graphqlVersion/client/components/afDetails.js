// @flow

/* eslint-disable react/no-array-index-key */

import React from 'react';
import { graphql } from 'react-relay';
import moment from 'moment';
import { flexItem, Icon, LargeMessage } from 'giu';
import _t from '../../translate';
import { COLORS } from '../gral/constants';
import QueryRendererWrapper from './uuQueryRendererWrapper';

// ==========================================
// Component declarations
// ==========================================
type PublicProps = {
  // lang: string,
  // Unit testing
  _now?: any, // eslint-disable-line
  selectedKeyId: ?string,
};

type Props = {
  ...$Exact<PublicProps>,
  // Relay
  node: ?Object,
};

const query = graphql`
  query afDetailsQuery($selectedKeyId: ID!) {
    node(id: $selectedKeyId) {
      ... on Key {
        firstUsed
        unusedSince
        description
        sources
        scope
      }
    }
  }
`;

// ==========================================
// Component
// ==========================================
class Details extends React.PureComponent {
  props: Props;
  state: { key: ?Object };

  constructor(props: Props) {
    super(props);
    this.state = {
      key: props.node,
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.node) this.setState({ key: nextProps.node });
  }

  render() {
    return (
      <div style={style.outer}>
        <div style={style.title}>
          {_t('msgDetailsView_Details').toUpperCase()}
        </div>
        {this.renderContents()}
      </div>
    );
  }

  renderContents() {
    if (this.props.selectedKeyId == null) {
      return (
        <LargeMessage>{_t('msgDetailsView_No message selected')}</LargeMessage>
      );
    }
    if (!this.state.key) {
      return (
        <LargeMessage>
          <Icon icon="circle-o-notch" />
        </LargeMessage>
      );
    }
    const {
      description,
      sources,
      firstUsed,
      unusedSince,
      scope,
    } = this.state.key;
    const since = this.renderDate(firstUsed);
    const until = unusedSince ? (
      <span>
        {' '}
        {_t('msgDetailsView_until')} {this.renderDate(unusedSince)}
      </span>
    ) : (
      ':'
    );
    const elSources = sources.length ? (
      <ul style={style.srcList}>
        {sources.map((src, idx) => (
          <li key={idx}>{src}</li>
        ))}
      </ul>
    ) : null;
    return (
      <div>
        {description && <div>{description}</div>}
        {scope && (
          <div>
            {_t('msgDetailsView_Scope')}:{' '}
            <span style={style.detailValue}>{scope}</span>
          </div>
        )}
        {_t('msgDetailsView_Used since')} {since}
        {until}
        {elSources}
      </div>
    );
  }

  renderDate(d: string) {
    const { _now } = this.props;
    return (
      <span title={moment(d).format('LLLL')} style={style.detailValue}>
        {_now ? moment(d).from(_now) : moment(d).fromNow()}
      </span>
    );
  }
}

// ------------------------------------------
const style = {
  outer: flexItem('none', {
    minHeight: 110,
    backgroundColor: COLORS.medium,
    padding: 5,
    marginTop: 5,
  }),
  title: {
    fontWeight: 900,
    letterSpacing: 3,
    textAlign: 'center',
    marginBottom: 10,
  },
  detailValue: {
    fontWeight: 'bold',
    color: '#444',
  },
  srcList: {
    marginTop: 0,
  },
};

// ==========================================
// Public API
// ==========================================
const Container = (props: PublicProps) => (
  <QueryRendererWrapper
    query={query}
    vars={{ selectedKeyId: props.selectedKeyId }}
    Component={Details}
    renderDuringLoad
    {...props}
  />
);

export default Container;
export { Details as _Details };
