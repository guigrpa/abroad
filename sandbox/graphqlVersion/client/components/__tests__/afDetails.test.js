/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { _Details as Details } from '../afDetails';

// https://github.com/facebook/react/issues/7386#issuecomment-238091398
jest.mock('react-dom');

// ======================================================
// Fixtures
// ======================================================
const KEY_ID = 'keyId';

const NODE_USED = {
  firstUsed: '2016-04-18T05:37:47.074Z',
  unusedSince: null,
  description: 'Some details',
  sources: ['src/client/components/060-translator.js'],
};

const NODE_UNUSED = {
  firstUsed: '2016-04-18T05:37:47.074Z',
  unusedSince: '2016-04-21T05:37:47.074Z',
  description: 'Some details',
  sources: [],
};

const NOW = new Date('2017-01-01T17:00:00.000Z');

// ======================================================
// Tests
// ======================================================
describe('Details', () => {
  it('01 no selection', () => {
    const tree = renderer
      .create(<Details node={null} selectedKeyId={null} _now={NOW} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('02 selection, but no data yet', () => {
    const tree = renderer
      .create(<Details node={null} selectedKeyId={KEY_ID} _now={NOW} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('03 selection (and data are available):', () => {
    it('03a message that is used', () => {
      const tree = renderer
        .create(<Details node={NODE_USED} selectedKeyId={KEY_ID} _now={NOW} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('03b message that is unused', () => {
      const tree = renderer
        .create(
          <Details node={NODE_UNUSED} selectedKeyId={KEY_ID} _now={NOW} />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
