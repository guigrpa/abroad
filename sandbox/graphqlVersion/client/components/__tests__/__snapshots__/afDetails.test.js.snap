// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details 01 no selection 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div
    className="giu-large-message"
    style={
      Object {
        "color": "#999999",
        "fontSize": "1.4em",
        "fontWeight": 700,
        "padding": "0.8em",
        "textAlign": "center",
      }
    }
  >
    No message selected
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div
    class="giu-large-message"
    style="color:#999999;font-size:1.4em;font-weight:700;padding:0.8em;text-align:center;"
  >
    No message selected
  </div>
</div>
`;

exports[`Details 02 selection, but no data yet 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div
    className="giu-large-message"
    style={
      Object {
        "color": "#999999",
        "fontSize": "1.4em",
        "fontWeight": 700,
        "padding": "0.8em",
        "textAlign": "center",
      }
    }
  >
    <i
      className="giu-icon fa fa-circle-o-notch fa-spin"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div
    class="giu-large-message"
    style="color:#999999;font-size:1.4em;font-weight:700;padding:0.8em;text-align:center;"
  >
    <i
      class="giu-icon fa fa-circle-o-notch fa-spin"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
</div>
`;

exports[`Details 03 selection (and data are available): 03a message that is used 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div>
    <div>
      Some details
    </div>
    Used since
     
    <span
      style={
        Object {
          "color": "#444",
          "fontWeight": "bold",
        }
      }
      title="Monday, April 18, 2016 7:37 AM"
    >
      8 months ago
    </span>
    :
    <ul
      style={
        Object {
          "marginTop": 0,
        }
      }
    >
      <li>
        src/client/components/060-translator.js
      </li>
    </ul>
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div>
    <div>
      Some details
    </div>
    Used since
     
    <span
      style="color:#444;font-weight:bold;"
      title="Monday, April 18, 2016 7:37 AM"
    >
      8 months ago
    </span>
    :
    <ul style="margin-top:0;">
      <li>
        src/client/components/060-translator.js
      </li>
    </ul>
  </div>
</div>
`;

exports[`Details 03 selection (and data are available): 03b message that is unused 1`] = `
<div
  style={
    Object {
      "WebkitFlex": "none",
      "backgroundColor": "#dcd6ff",
      "flex": "none",
      "marginTop": 5,
      "minHeight": 110,
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": 900,
        "letterSpacing": 3,
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    DETAILS
  </div>
  <div>
    <div>
      Some details
    </div>
    Used since
     
    <span
      style={
        Object {
          "color": "#444",
          "fontWeight": "bold",
        }
      }
      title="Monday, April 18, 2016 7:37 AM"
    >
      8 months ago
    </span>
    <span>
       
      until
       
      <span
        style={
          Object {
            "color": "#444",
            "fontWeight": "bold",
          }
        }
        title="Thursday, April 21, 2016 7:37 AM"
      >
        8 months ago
      </span>
    </span>
  </div>
</div>
------------HTML PREVIEW---------------
<div style="-webkit-flex:none;background-color:#dcd6ff;flex:none;margin-top:5px;min-height:110px;padding:5px;">
  <div style="font-weight:900;letter-spacing:3px;margin-bottom:10px;text-align:center;">
    DETAILS
  </div>
  <div>
    <div>
      Some details
    </div>
    Used since
     
    <span
      style="color:#444;font-weight:bold;"
      title="Monday, April 18, 2016 7:37 AM"
    >
      8 months ago
    </span>
    <span>
       
      until
       
      <span
        style="color:#444;font-weight:bold;"
        title="Thursday, April 21, 2016 7:37 AM"
      >
        8 months ago
      </span>
    </span>
  </div>
</div>
`;
